sandboxnu/cooper
# User Onboarding Flow

## Description

Adds a user onboarding flow to _force_ the user to create a profile.

**Please read this before reviewing:**
- Please, please, please 🎵 DESTROY this PR in code review
- The onboarding flow can be initiated by wrapping any layout with the `OnboardingWrapper` component
    - I have to do some more research to make sure that this doesn't cause re-render hell / we don't fetch the user profile every second
    - Future Rishi: yes I was right there is a re-render problem caused by fetching companies and roles ...
- Currently, the "Take me there!" button after finishing onboarding (and saying that you have finished a co-op) does not work. It only closes the dialog for now but ideal behavior is to close the dialog _and_ open the "+ New Review" dialog.
    - Will have to think about this one for a bit since I couldn't find an elegant way of triggering that (it sits in the header).
- The design is _almost_ the same as the one on the [Figma](https://www.figma.com/design/YPT3Wi0EVi9rxYmObF6cC4/UI-Designs?node-id=1-5&t=pWN0Yda0T5QSj7Dk-0) 
- _except_ the Yes/No field for previous co-ops. I will open a separate PR for that but this onboarding flow should be functional.
  - Tailwind lover banushi-a fixed this <3
  - I also made it responsive in the meantime
- Dialog closes and doesn't save state when you browse to a different page / switch windows.

**For Testing:**
- Testing using the Vercel Preview
    - Delete any user profile associated with your account from the "production" database (on Vercel)
    - Go to the latest Vercel Preview
    - Sign In using the button on the top right
    - Go to `/roles` / `/companies` and confirm that the onboarding flow is initiated
    - Finish the onboarding flow in both cases: said "Yes" to previous co-ops and said "No" to previous co-ops.
- Testing locally
    - Same as above but you'll (hopefully 🤨) be using a development Neon database
    - Delete any profiles from this database (Use Drizzle studio using `pnpm db:studio`
    - I would recommend doing a `pnpm build` instead of `pnpm dev` since it will take a hot sec for the onboarding flow to appear if you run it in dev mode.



## Motivation and Context



Closes #88

## How has this been tested?





With me eyes (no but fr will add frontend unit and e2e tests after showcase)

## Screenshots (if appropriate):

**Signed In User w/ No Profile**
![image](https://github.com/user-attachments/assets/b2966b5d-37a5-4e18-bd19-24fb2f035b79)

**Post Onboarding Prompt - Has done a co-op before**
![image](https://github.com/user-attachments/assets/e3a7b9ec-9cbb-4b9a-9dfb-3a31983e5019)

**Post Onboarding Prompt - Has not done a co-op before**
![image](https://github.com/user-attachments/assets/bc253a6d-39e8-46df-8d17-6e289f211b25)

## Types of changes




## Checklist:




- [x] I have moved the ticket to "In Review"
- [x] I have added reviewers to this PR
